Cannot run 'sensei pdf' on CircleCI

Console output:
```
Generate pdf slides & labs
Build slides & labs
Training material folder: '/home/circleci/project'. This can be changed using '--env.material=<relative path to training material folder>'.
HTML files generation failed! null
(node:207) UnhandledPromiseRejectionWarning: undefined
(Use `node --trace-warnings ...` to show where the warning was created)
(node:207) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:207) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
CircleCI received exit code 0
```

Note that the exit code is 0 even though the PDFs were not generated.

See https://app.circleci.com/pipelines/github/Zenika/formation-deeplearning-tensorflow2/17/workflows/c6aaf39c-8fe8-409c-b9b8-48ad2560cdfa/jobs/37

Error is now fully logged.

Console output:

```

ERROR in ./Slides/04_2_CNN.md?chapterIndex=8
Module not found: Error: Can't resolve './ressources/CNN/lenet_archi.png' in '/home/circleci/project/Slides'
 @ ./Slides/04_2_CNN.md?chapterIndex=8 16:34-77
 @ ./Slides/slides.json
 @ ./node_modules/@zenika/sensei/src/slides/slides.js

ERROR in ./node_modules/@zenika/sensei/src/labs/labs.js
Module not found: Error: Can't resolve 'prismjs/themes/prism.css' in '/home/circleci/project/node_modules/@zenika/sensei/src/labs'
 @ ./node_modules/@zenika/sensei/src/labs/labs.js 1:0-34

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'prismjs/themes/prism.css' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 6:0-34

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'reveal.js/dist/reveal.css' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 5:0-35

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'reveal.js/dist/reveal.js' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 1:0-46 13:0-6

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'reveal.js/plugin/math/math.js' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 4:0-64 25:12-31

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'reveal.js/plugin/notes/notes.js' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 2:0-64 25:33-50

ERROR in ./node_modules/@zenika/sensei/src/slides/slides.js
Module not found: Error: Can't resolve 'reveal.js/plugin/zoom/zoom.js' in '/home/circleci/project/node_modules/@zenika/sensei/src/slides'
 @ ./node_modules/@zenika/sensei/src/slides/slides.js 3:0-61 25:52-68
```

See https://app.circleci.com/pipelines/github/Zenika/formation-deeplearning-tensorflow2/19/workflows/f7132815-f10f-4952-8b94-da8f688db754/jobs/42

--- a/bin/sensei.js
+++ b/bin/sensei.js
@@ -1,4 +1,5 @@
 #! /usr/bin/env node
+
 const spawn = require("child_process").spawn;
 const fork = require("child_process").fork;
 const path = require("path");
@@ -6,17 +7,17 @@ const WebpackDevServer = require("webpack-dev-server");
 const config = require("../webpack.config");
 const webpack = require("webpack");
 
-function cli(args) {
+async function cli(args) {
   let trainingMaterialFolder = args[3];
   switch (args[2]) {
     case "serve":
       serve(trainingMaterialFolder);
       break;
     case "build":
-      build(trainingMaterialFolder);
+      await build(trainingMaterialFolder);
       break;
     case "pdf":
-      pdf(trainingMaterialFolder);
+      await pdf(trainingMaterialFolder);
       break;
     case "help":
     case "-h":
@@ -150,6 +151,11 @@ process.on("SIGTERM", function onSigterm() {
     new Date().toISOString()
   );
   process.exit();
+});
+
+process.on("unhandledRejection", (err) => {
+  process.exitCode = 1;
+  throw err;
 });
 
 cli(process.argv);

--- a/bin/sensei.js
+++ b/bin/sensei.js
@@ -46,13 +46,14 @@ function build(trainingMaterialFolder = ".") {
   console.log("Build slides & labs");
   return new Promise((resolve, reject) => {
     webpack(config({ material: trainingMaterialFolder }), (err, stats) => {
-      if (err || stats.hasErrors()) {
-        console.error("HTML files generation failed!", err);
-        reject();
-      } else {
-        console.log("Files generated to dist folder");
-        resolve();
+      if (err) {
+        return reject(err);
       }
+      if (stats.hasErrors()) {
+        return reject(stats);
+      }
+      console.log("Files generated to dist folder");
+      resolve();
     });
   });
 }

--- a/bin/sensei.js
+++ b/bin/sensei.js
@@ -101,7 +101,12 @@ function pdfSlides(trainingName) {
       process.stdout.write(data);
     });
 
-    child.on("exit", function () {
+    child.on("exit", function (code) {
+      if (code !== 0) {
+        return reject(
+          new Error(`spawned process exited with non-zero code '${code}'`)
+        );
+      }
       console.log("PDF slides generated");
       resolve();
     });
@@ -123,7 +128,12 @@ function pdfLabs(trainingName) {
       ]
     );
 
-    child.on("exit", function () {
+    child.on("exit", function (code) {
+      if (code !== 0) {
+        return reject(
+          new Error(`forked process exited with non-zero code '${code}'`)
+        );
+      }
       console.log("PDF labs generated");
       resolve();
     });

--- a/bin/sensei.js
+++ b/bin/sensei.js
@@ -50,7 +50,7 @@ function build(trainingMaterialFolder = ".") {
         return reject(err);
       }
       if (stats.hasErrors()) {
-        return reject(stats);
+        return reject(new Error(stats.toString()));
       }
       console.log("Files generated to dist folder");
       resolve();

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -74,6 +74,14 @@ module.exports = (env = {}, argv) => {
       alias: {
         "training-material": trainingMaterialFolder,
       },
+      symlinks: false,
+      modules: [
+        // this is to support the "npm link" case where imports must be resolved
+        // from the project folder
+        path.resolve(__dirname, "node_modules"),
+        // this is the default and works for most cases
+        "node_modules"
+      ],
     },
     output: {
       path: path.resolve("./dist"),

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -75,7 +75,7 @@ module.exports = (env = {}, argv) => {
         "training-material": trainingMaterialFolder,
       },
       symlinks: false,
-      modules: [path.resolve(__dirname, "node_modules"), path.resolve("./")],
+      modules: [path.resolve("./")],
     },
     output: {
       path: path.resolve("./dist"),

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -80,7 +80,7 @@ module.exports = (env = {}, argv) => {
         // from the project folder
         path.resolve(__dirname, "node_modules"),
         // this is the default and works for most cases
-        "node_modules"
+        "node_modules",
       ],
     },
     output: {

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -70,13 +70,6 @@ module.exports = (env = {}, argv) => {
         },
       ],
     },
-    resolve: {
-      alias: {
-        "training-material": trainingMaterialFolder,
-      },
-      symlinks: false,
-      modules: [path.resolve("./")],
-    },
     output: {
       path: path.resolve("./dist"),
     },

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -70,6 +70,11 @@ module.exports = (env = {}, argv) => {
         },
       ],
     },
+    resolve: {
+      alias: {
+        "training-material": trainingMaterialFolder,
+      },
+    },
     output: {
       path: path.resolve("./dist"),
     },

--- a/webpack.config.js
+++ b/webpack.config.js
@@ -75,7 +75,7 @@ module.exports = (env = {}, argv) => {
         "training-material": trainingMaterialFolder,
       },
       symlinks: false,
-      modules: [path.resolve(__dirname, "node_modules")],
+      modules: [path.resolve(__dirname, "node_modules"), path.resolve("./")],
     },
     output: {
       path: path.resolve("./dist"),

